### PR TITLE
Force kill takover failure proc

### PR DESCRIPTION
**Pull request by HJianBo**

### CODE REVIEW

File: src/emqx_cm.erl [hunk 1]
```diff
@@ -286,44 +295,63 @@ discard_session(ClientId) when is_binary(ClientId) ->
...
                   reason => Error,
                   stacktrace => St,
                   stale_channel => stale_channel_info(Pid)
                  }),
-            ok = force_kill(Pid)
+            ok = force_kill(Pid),
+            return_error_if_action_is_takeover(Action, Error)
     end.
 
+return_error_if_action_is_takeover({takeover, _}, Reason) ->
```
Comment (DDDHuang) [inline src/emqx_cm.erl:350]: Maybe a batter func name. @zmstone Any suggestions?

**Review by DDDHuang — DISMISSED** [review]

File: apps/emqx_exproto/src/emqx_exproto.appup.src [hunk 2]
```diff
@@ -1,28 +1,22 @@
...
     [{load_module,emqx_exproto_channel,brutal_purge,soft_purge,[]}]},
-   {"4.3.3",
+   {<<"4\\.3\\.[23]">>,
     [{load_module,emqx_exproto_conn,brutal_purge,soft_purge,[]},
      {load_module,emqx_exproto_channel,brutal_purge,soft_purge,[]}]},
-   {"4.3.2",
-    [{load_module,emqx_exproto_conn,brutal_purge,soft_purge,[]},
-     {load_module,emqx_exproto_channel,brutal_purge,soft_purge,[]}]},
-   {<<"4.3.[0-1]">>,
+   {<<"4\\.3\\.[01]">>,
```
Comment (zmstone) [inline apps/emqx_exproto/src/emqx_exproto.appup.src:8]:
```suggestion
   {<<"4\\.3\\.[0-1]">>,
```

Comment (zmstone) [reply apps/emqx_exproto/src/emqx_exproto.appup.src:8]: keep it the same as before

File: apps/emqx_exproto/src/emqx_exproto.appup.src [hunk 3]
```diff
@@ -1,28 +1,22 @@
...
     [{load_module,emqx_exproto_gsvr,brutal_purge,soft_purge,[]},
      {load_module,emqx_exproto_gcli,brutal_purge,soft_purge,[]},
      {load_module,emqx_exproto_conn,brutal_purge,soft_purge,[]},
      {load_module,emqx_exproto_channel,brutal_purge,soft_purge,[]}]},
    {<<".*">>,[]}],
-  [{"4.3.4",
+  [{<<"4\\.3\\.[45]">>,
```
Comment (zmstone) [inline apps/emqx_exproto/src/emqx_exproto.appup.src:14]:
```suggestion
  [{<<"4\\.3\\.[4-5]">>,
```
To make it more inclusive when add a 6 next time

File: src/emqx_cm.erl [hunk 4]
```diff
@@ -226,18 +226,24 @@ open_session(true, ClientInfo = #{clientid := ClientId}, ConnInfo) ->
...
                               Session = create_session(ClientInfo, ConnInfo),
                               register_channel(ClientId, Self, ConnInfo),
                               {ok, #{session => Session, present => false}}
+                          end,
+                      case takeover_session(ClientId) of
+                          {ok, ConnMod, ChanPid, Session} ->
+                              ok = emqx_session:resume(ClientInfo, Session),
+                              case call_or_kill({takeover, 'end'}, ConnMod, ChanPid) of
+                                  {error, _} -> CreateSess();
+                                  Pendings ->
```
Comment (zmstone) [inline src/emqx_cm.erl:240]:
```suggestion
                                  {ok, Pendings} ->
```
And ok clause should go first, (before error)

File: src/emqx_cm.erl [hunk 5]
```diff
@@ -286,44 +295,63 @@ discard_session(ClientId) when is_binary(ClientId) ->
...
 %% Keeping the stale pid around, or returning error or raise an exception
 %% benefits nobody.
--spec kick_or_kill(kick | discard, module(), pid()) -> ok.
-kick_or_kill(Action, ConnMod, Pid) ->
+-spec call_or_kill(Action, module(), pid())
+    -> ok                        %% returned by kick, discard
+     | emqx_session:session()    %% returned by {takeover, 'begin'}
+     | list(emqx_type:deliver()) %% returned by {takeover, 'end'}
+  when Action :: kick | discard | {takeover, 'begin'} | {takeover, 'end'}.
+call_or_kill(Action, ConnMod, Pid) ->
```
Comment (zmstone) [inline src/emqx_cm.erl:307]: Maybe just name it ‘takeover’

Kick, kill are actually a part of the takeover process.
The mfa reported in the logs would be more clear tooo

File: src/emqx_cm.erl [hunk 6]
```diff
@@ -286,44 +295,63 @@ discard_session(ClientId) when is_binary(ClientId) ->
...
 
-%% @private Kick a local stale session to force it step down.
-%% If failed to kick (e.g. timeout) force a kill.
+%% @private call a local stale session to execute an Action.
+%% If failed to response (e.g. timeout) force a kill.
 %% Keeping the stale pid around, or returning error or raise an exception
 %% benefits nobody.
--spec kick_or_kill(kick | discard, module(), pid()) -> ok.
-kick_or_kill(Action, ConnMod, Pid) ->
+-spec call_or_kill(Action, module(), pid())
```
Comment (zmstone) [inline src/emqx_cm.erl:302]: It also returns {error, …}

File: src/emqx_cm.erl (hunk 1 above)
Comment (zmstone) [reply src/emqx_cm.erl:350]: maybe no need for this function if refactor as suggested above

**Review by terry-xiaoyu — DISMISSED** [review]

**Review by terry-xiaoyu — APPROVED** [review]

**Pull request merged**